airflow/contrib/operators/mysql_to_gcs.py

```diff
@@ -135,26 +182,83 @@ def _write_local_data_files(self, cursor):
         tmp_file_handles = {self.filename.format(file_no): tmp_file_handle}
 
+        # Save file header for csv if required
+        if(self.export_format['file_format'] == 'csv'):
+
+            # Deal with CSV formatting. Try to use dialect if passed
+            if('csv_dialect' in self.export_format):
+                # Use dialect name from params
+                dialect_name = self.export_format['csv_dialect']
+            else:
+                # Create internal dialect based on parameters passed
+                dialect_name = 'mysql_to_gcs'
+                csv.register_dialect(dialect_name,
+                                     delimiter=self.export_format.get('csv_delimiter') or
+                                     ',',
+                                     doublequote=self.export_format.get(
+                                         'csv_doublequote') or
+                                     'True',
+                                     escapechar=self.export_format.get(
+                                         'csv_escapechar') or
+                                     None,
+                                     lineterminator=self.export_format.get(
+                                         'csv_lineterminator') or
+                                     '\r\n',
+                                     quotechar=self.export_format.get('csv_quotechar') or
+                                     '"',
+                                     quoting=eval(self.export_format.get(
+                                         'csv_quoting') or
+                                         'csv.QUOTE_MINIMAL'))
+            # Create CSV writer using either provided or generated dialect
+            csv_writer = csv.writer(tmp_file_handle,
+                                    encoding='utf-8',
+                                    dialect=dialect_name)
+
+            # Include column header in first row
+            if('csv_columnheader' in self.export_format and
+                    eval(self.export_format['csv_columnheader'])):
+                csv_writer.writerow(schema)
+
         for row in cursor:
-            # Convert datetime objects to utc seconds, and decimals to floats
+
+            # Convert datetimes and longs to BigQuery safe types
             row = map(self.convert_types, row)
-            row_dict = dict(zip(schema, row))
 
-            # TODO validate that row isn't > 2MB. BQ enforces a hard row size of 2MB.
-            s = json.dumps(row_dict)
-            if PY3:
-                s = s.encode('utf-8')
-            tmp_file_handle.write(s)
+            # Save rows as JSON
+            if(self.export_format['file_format'] == 'json'):
+                # Convert datetime objects to utc seconds, and decimals to floats
+
+                row_dict = dict(zip(schema, row))
+
+                # TODO validate that row isn't > 2MB. BQ enforces a hard row size of 2MB.
+                s = json.dumps(row_dict, sort_keys=True)
```

Comment:
What is the performance penalty of this operator? We only need it when we are dumping csv.

Reply:
I can leave 'json' as the else option for the same 'if' within the loop, great point. I initially thought to have two separate loops (one for 'csv' and one for 'json'), maybe that's still the best.